tools/shared/src/main/scala/org/scalajs/core/tools/linker/backend/emitter/ScalaJSClassEmitter.scala

```diff
@@ -72,6 +74,73 @@ private[scalajs] final class ScalaJSClassEmitter(
 
   private[emitter] def semantics: Semantics = linkingUnit.semantics
 
+  private[emitter] lazy val typeTags: Map[LinkedClass, Int] = {
+    import Definitions.ReservedTags
+
+    val childrenMap =
+      linkingUnit.classDefs.groupBy(_.superClass.map(_.name)).withDefaultValue(Nil)
+
+    // Some classes may have parents that have been removed
+    val orphans = for(cls <- linkingUnit.classDefs;
+                    parent <- cls.superClass
+                    if !linkedClassByName.isDefinedAt(parent.name)) yield cls
+
+    val roots = childrenMap(None) ++ orphans
+
+    @tailrec def getTypeTags(curTag: Int, stack: Seq[LinkedClass],
+      accu: Map[LinkedClass, Int] = Map.empty): Map[LinkedClass, Int] = stack match {
+      case cur +: rest if ReservedTags.isDefinedAt(cur.name.name) =>
+        val children = childrenMap(Some(cur.name.name))
+        val tag = ReservedTags(cur.name.name)
+        getTypeTags(curTag, children ++ rest, accu + (cur -> tag))
+
+      case cur +: rest =>
+        val nextTag = curTag + 1
+        val children = childrenMap(Some(cur.name.name))
+        getTypeTags(nextTag, children ++ rest, accu + (cur -> curTag))
+
+      case Nil =>
+        // Sanity check: all classDefs must have a tag
+        assert(linkingUnit.classDefs.forall(accu.isDefinedAt))
+        accu
+    }
```

Comment:
On a different topic, this algorithm is not *stable*, in the sense that nothing guarantees that, on two consecutive runs, it will give the same tags to the same classes. This could be annoying as compiling the same codebase twice does not give the same .js file.

It would be good to stabilize it, for example by sorting `roots` and `children` by `_.name.name`.